features/prog.go

```diff
@@ -267,30 +263,65 @@ func haveProgramHelper(pt ebpf.ProgramType, helper asm.BuiltinFunc) error {
 	}
 
 	prog, err := ebpf.NewProgramWithOptions(spec, ebpf.ProgramOptions{
-		LogDisabled: true,
+		LogLevel: 1,
 	})
 	if err == nil {
 		prog.Close()
 	}
 
+	var verr *ebpf.VerifierError
+	if !errors.As(err, &verr) {
+		return err
+	}
+
+	helperTag := fmt.Sprintf("#%d", helper)
+
 	switch {
 	// EACCES occurs when attempting to create a program probe with a helper
 	// while the register args when calling this helper aren't set up properly.
 	// We interpret this as the helper being available, because the verifier
 	// returns EINVAL if the helper is not supported by the running kernel.
 	case errors.Is(err, unix.EACCES):
-		// TODO: possibly we need to check verifier output here to be sure
 		err = nil
 
 	// EINVAL occurs when attempting to create a program with an unknown helper.
 	case errors.Is(err, unix.EINVAL):
-		// TODO: possibly we need to check verifier output here to be sure
-		err = ebpf.ErrNotSupported
+		// https://github.com/torvalds/linux/blob/09a0fa92e5b45e99cf435b2fbf5ebcf889cf8780/kernel/bpf/verifier.c#L10663
+		if logContainsAll(verr.Log, "invalid func", helperTag) {
+			return ebpf.ErrNotSupported
+		}
+
+		// https://github.com/torvalds/linux/blob/09a0fa92e5b45e99cf435b2fbf5ebcf889cf8780/kernel/bpf/verifier.c#L10668
+		if logContainsAll(verr.Log, "program of this type cannot use helper", helperTag) {
+			return ebpf.ErrNotSupported
```

Comment:
I think it would help users to distinguish between the different kinds of `ErrNotSupported`. So maybe extend it with something like this?
```suggestion
			return fmt.Errorf("program of this type cannot use helper: %w", ebpf.ErrNotSupported)
```